### ISSUE TITLE
Several Interface changes need to be accounted for with Entity Framework Core 3.0 and .NET Core 3.0

### ISSUE THREAD

**Issue by NathanFrasier** [romantitov/MockQueryable#17]
Using EntityFrameworkCore 3.0.0-preview7.19362.6 The `IAsyncEnumerable` `IAsyncQueryProvider` and `IAsyncEnumerator` interfaces require different members.

As far as I can tell the member changes are as follows:

`IAsyncEnumerable`:
* Adds `GetAsyncEnumerator(CancellationToken)`

`IAsyncQueryProvider`:
* Adds `ExecuteAsync<TResult>(Expression, CancellationToken)` with return type of `TResult`

`IAsyncEnumerator`:
* Adds `MoveNextAsync()`

`IAsyncDisposable`:
* Adds `DisposeAsync()`

As far as I can tell, the `ExecuteAsync()` methods that return `IAsyncEnumerable` or `Task<TResult>` are unnecessary now.

These interface changes also appear to break some of the non-core packages as well, apparrently due to the removal of the `GetEnumerator` method on `IAsyncEnumerable` (I believe this just needs to be changed to `GetAsyncEnumerator` instead)

related GitHub issue:
https://github.com/aspnet/EntityFrameworkCore/issues/12047

**Comment by V0ldek** [comment]
Hello,
I've been using this library to test my repos and recently I was migrating them to EF Core 3.0 preview 7. That of course broke the library, so I had to hack some code by myself. I can chime in on how I resolved the issues:

`IAsyncEnumerable`:
- Existing `IAsyncEnumerable<T> TestAsyncEnumerable.GetEnumerator()` needs to have its signature changed to `GetAsyncEnumerator(CancellationToken cancellationToken = default)`. The behavior remains the same.

`IAsyncEnumerator`, `IAsyncDisposable`:
- Also straightforward changes
```
ValueTask<bool> MoveNextAsync() => new ValueTask<bool>(_enumerator.MoveNext());

public ValueTask DisposeAsync()
{
    _source.Dispose();
    return new ValueTask();
}
```

`IAsyncQueryProvider`:
- Now `ExecuteAsync` is a tricky one. The EF Core team specifically said that they have no intention of supporting fake async calls (see [this issue](https://github.com/aspnet/EntityFrameworkCore/issues/16864)). The suggested solution is to use reflection on `TResult` to find out what needs to be returned.

```
public TResult ExecuteAsync<TResult>(
    Expression expression,
    CancellationToken cancellationToken = new CancellationToken())
{
    var expectedResultType = typeof(TResult).GetGenericArguments()[0];
    var executionResult = ((IQueryProvider) this).Execute(expression);

    return (TResult) typeof(Task).GetMethod(nameof(Task.FromResult))
        .MakeGenericMethod(expectedResultType)
        .Invoke(null, new[] {executionResult});
}
```

I didn't find an EF call where `TResult` wasn't a `Task<TInner>` or the `TInner` was different from the Expression's type, but these are possible causes for binding exceptions. To get more sensible error messages checks for `typeof(Task).IsAssignableFrom(typeof(TResult))` and `expectedResultType.IsAssignableFrom(expression.Type)`. 

Either way, the `IAsyncQueryProvider` interface is marked as volatile, and "may be changed or removed without notice in any release", so all of this is applicable to preview 7 and might become obsolete in preview 8.

If this looks sensible I could put my solution into a temporary branch and test it against preview 7 assemblies. You'll also free to take what I wrote and do whatever you deem appropriate for this lib.

**Comment by NathanFrasier** [comment]
I had Intentions to create a PR but this work seems more thorough than what I was able to put together, specifically with regard to the `ExecuteAsync` call. I wasn't expecting the issue to be completely resolved until EF Core 3.0 hits stable. Since you already have a more thorough solution you can make the PR, otherwise I'll combine what you've looked into with my code and make one when I have some spare time.

**Comment by V0ldek** [comment]
So I added the code as a PR so everyone can look at it, but I don't think it should be merged before EF Core 3.0 stable release.

Fortunately the linked CI service uses .NET Core 2.2 SDK, so that check is not going to pass anyway :)

**Comment by romantitov** [comment]
Thanks @NathanFrasier  and @V0ldek  MockQueryable from version 3.0.0 supports EF Core 3
